---
title: 'A pipeline for detecting HIV / AIDS in rural populations'
author: Philip Bester and Andrie de Vries
date: '2019-03-06'
slug: aids-hiv-detection
categories:
  - R Language
  - Guest Post
  - R in Medicine
tags: [Medicine, R/Medicine, HIV/AIDS]
summary: 'How we created an affordable pipeline with Python and R running on a Raspberry Pi to detect HIV/AIDS in rural populations in South Africa'

---

```{r, include=FALSE}
knitr::opts_chunk$set(eval = TRUE)

# ensure packages are installed
lapply(c('tibble'), function(x) {
  if (!requireNamespace(x)) install.packages(x)
})

library(tibble)
```   




It is because I am not a graph analysis expert that I thought it important to write this article. For someone who thinks in terms of single rectangular data sets, it is a bit of a mental leap to understand how to apply *tidy* principles to a more robust object, such as a graph table.  Thankfully, there are two packages that make this work much easier:

- [`tidygraph`](https://github.com/thomasp85/tidygraph) - Provides a way for `dplyr` to interact with graphs

- [`ggraph`](https://github.com/thomasp85/ggraph) - Extension to `ggplot2` for graph analysis

## Introduction

Research regarding the use of a Raspberry Pi, python, R and various other software was recently published in [`PLoS ONE`](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0213241).  The authors described how they used this cheap hardware to create a [`phylogenetic`](https://en.wikipedia.org/wiki/Phylogenetics) pipeline, tailored for the HIV drug resistance testing facility.  
As with antibiotics, the human immunodeficiency virus (HIV) can develop resistance to the life-saving anti-retroviral (ARV) drugs.  To an extent, this was dramatized in the movie, [`Dallas Buyers Club`](https://en.wikipedia.org/wiki/Dallas_Buyers_Club), where an American electrician was diagnosed with HIV, at a time when treatment options were essentially non-existent. He reads about AZT (Zidovudine) which at that stage was not yet FDA approved, and having been told that he has only a month or two to live,  decides to smuggle this illegal drug across the Mexican border.  Although AZT is still a very effective drug today, we have come to realise that using just one drug or drugs of only one class, is ineffective in the long term, due to the development of resistance.  Today, we use HAART (highly active antiretroviral treatment) or cART (combination antiretroviral treatment) which typically consists of 3 drugs from 2 different classes.  This way, if replication derived mutations do result in a mutation conveying resistance against one of the drugs, the other two drugs are there to suppress this virion and it cannot escape.
Nowadays, health care professionals have many ARVs to choose from and knowing which drugs will suppress the viral replication in an individual patient, is very valuable in making an informed choice.  The routine way of accomplishing this is to sequence specific areas of the HIV genome which were isolated from the patient.  Analysis of this sequence data is used to detect mutations which convey resistance to specific drugs.  HIV has a very high mutation rate, which is the mechanism for developing drug resistance, but this is also used to our advantage in the drug resistance testing facility.  During the extraction of nucleic acid from the patient sample, the viral RNA is isolated via reverse transcription and polymerase chain reaction.  This isolated and amplified nucleic acid can then be subjected to [`Sanger sequencing`](https://en.wikipedia.org/wiki/Sanger_sequencing) or other sequencing methods. Importantly, the polymerase chain reaction is very sensitive to contamination of other non-subject nucleic products, which means that cross-contamination can create an erroneous result.  In addition, the process is very labour intensive and mistakes such as sample swaps can easily occur.  For these reasons, it is common practise to do phylogenetic analysis on sequence results in order to detect any anomalies. The process of phylogenetic inference involves finding all the important sequences to include in the analysis, aligning these sequences, curating this alignment, running the phylogenetic model parameter estimation calculations and finally rendering it.  This can be a very involved process.  The researchers in this publication developed a portable, self-contained and affordable solution to this problem.  This pipeline can detect previously sampled sequences out of a self-maintaining database to include in the analysis, which is then displayed as a [`phylogenetic tree`](https://en.wikipedia.org/wiki/Phylogenetic_tree) and a heat map of [`genetic distances`](https://en.wikipedia.org/wiki/Genetic_distance), using R packages.  
Before getting to the colouring of a heat map, it would be very useful to know how to colour a heat map to best draw attention to very similar viral sequences which can be further investigated.  The authors used publicly available sequences from the [`Los Alamos HIV sequence database`](https://www.hiv.lanl.gov/content/sequence/HIV/mainpage.html).  For subtypes B and C, using unique patient identifiers from the database, the sequences data was split into one file containing just one sequence per patient (to analyse type inter) and multiple files per patient, for patients with more than one sequence (to analyse type intra).  The inter type files were aligned as a whole using [`MAFFT`](https://mafft.cbrc.jp/alignment/software/) and each of the type intra files where aligned similarly, but individually.  This allowed them to calculate the genetic distances of viral sequences both between patients and within patients, i.e. type inter and type intra.


## Logistic regression

Provide a link to alignment data to use for the heatmap.
Go through the steps

## Heatmap

Provide a link to alignment data to use in the logistic regression
Go through the steps



