---
title: "Heatmap and tree"
author: "Armand"
date: "09 March 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Displaying genetic distances in a heatmap and relatedness in a phylogenetic tree

Load some packages

```{r warning=FALSE, message=FALSE}
library(ape)
library(tibble)
library(tidyr)
library(dplyr)
library(plotly)
library(RColorBrewer)
```

Read in the multiple sequence alingment file

```{r}
# Read in the alignment file
aln <- read.dna('example.aln', format = 'fasta')

# Calculate the genetic distances between sequences using the K80 model, as.mattrix makes the rest easier
alnDist <- dist.dna(aln, model = "K80", as.matrix = TRUE)
```




### Reduction of heatmap to focus on the important data

The pipeline mentioned, uses the Basic Local Alignment Search Tool (BLAST) to retrieve previously sampled sequences and adds these retrieved sequences to the analysis.  [BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi) is like a search engine you use on the web, but for protein or DNA sequences.  By doing this important sequences from retrospective samples are included which enables PhyloPi to be aware of past sequences and not just batch per batch aware.  Have a look at the [paper](https://journals.plos.org/plosone/article/comments?id=10.1371/journal.pone.0213241) for some examples.


The data we have, is ready to use for heatmap plotting purposes, but since the data we have here also contains previously sample sequences, comparing those sequences amongst themselfs would be a distraction.  We are interested in those samples, but only compare to the current batch of samples analysed.  The figures below should explain this a bit better. 


![Lots of destracting data](heatmap_full.png)



From the image above you can see that, typical of a heatmap, it is symetrical on the diagonal vertice.  We are showing submitted *vs* retrieved samples in both the horisontal and vertical direction.  Notice also, as annotated as "Distractio", are the previous samples compared amongst themselfs.  We are not interested in those sample now as we would allready have acted on any issues then.  What we rather want, is a heatmap as depicted in the image below.



![Data deserving our attention](heatmap_focused.png)



Luckily for us, we have a very powerful tool to our desposal, **R**, and plenty of really usefull and convenient packages, like dplyr.



```{r}
# alnDist[upper.tri(alnDist, diag = FALSE)] <- NA

alnDistLong <- 
  alnDist %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "sample_1") %>% 
  gather(key = "sample_2", value = "distance", -sample_1, na.rm = TRUE) %>% 
  arrange(distance)

alnDistLong
```




Create a new variable, combined, we will paste the names for sample1 and sample2 together


Final cleanup and removal of destracting data

```{r}
# get the names of samples originally in the fasta file used for submission
qSample <- names(read.dna("example.fasta", format = "fasta"))



focus.df <- 
  alnDistLong %>% 
  filter(
    sample_1 %in% qSample,
    sample_1 != sample_2
    )


```

Plot the heatmap using plotly for inteactivity

```{r, fig.width=10, fig.height=6}
focus.df %>% 
  plot_ly(
    x = ~sample_2,
    y = ~sample_1,
    z = ~distance,
    type = "heatmap", colors = brewer.pal(11, "RdYlGn"), 
    zmin = 0.0, zmax = 0.03,  xgap = 2, ygap = 1
) %>% 
  layout(
    margin = list(l = 100, r = 10, b = 100, t = 10, pad = 4), 
    yaxis = list(tickfont = list(size = 10), showspikes = TRUE),
    xaxis = list(tickfont = list(size = 10), showspikes = TRUE)
  )
```




## Phylogenetic tree

Above we have used the package [Ape](http://ape-package.ird.fr/) to calculate the genetic distances for the heatmap.  Another way of looking at our alingment data is to use phylogenetic inference. The PhyloPi pipeline saves each step of phylogenetic inference to allow the user to intercept at any step.  We can use the newick tree file (a text file formatted as newick) and draw our own tree.

```{r, fig.height=12, fig.width=7}
tree <- read.tree("example-tree.txt")

plot.phylo(
  tree, cex = 0.8, 
  use.edge.length = TRUE, 
  tip.color = 'blue', 
  align.tip.label = TRUE, 
  show.node.label = TRUE
)

```










